### PR TITLE
Avoid crash with monitor feature on Ruby 3.0 

### PR DESCRIPTION
**Pull request by Watson1978**
Seems that ImageMagick might call back in a different thread than Ruby is running in.
If it is a different thread, it would not have a Ruby GVL and it could not retrieve properly Ruby stack.

Unfortunately, there is no API available to check if the current thread has a GVL, so the thread id was checked in here.

There is a private API to check whether the current thread has a GVL :(
https://github.com/ruby/ruby/blob/06b44f819eb7b5ede1ff69cecb25682b56a1d60c/thread.c#L1908-L1919

Related to https://github.com/rmagick/rmagick/issues/464

### CODE REVIEW

**Comment by Watson1978** [comment]
`rm_progress_monitor()` causes SEGV with Ruby 3.0 on Linux.

```
[BUG] Segmentation fault at 0x00000000000000a0
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]

-- Machine register context ------------------------------------------------
 RIP: 0x000055baa09ec7f2 RBP: 0x00007fd9f41e1cf0 RSP: 0x00007fd9f41e1cd8
 RAX: 0x0000000000000000 RBX: 0x000055baa6b2bb50 RCX: 0x000055baa4b4ed80
 RDX: 0x00007fd9f41e1cd8 RDI: 0x0000000000000000 RSI: 0x0000000000000001
  R8: 0x000055baa6b2bdb8  R9: 0x0000000000000000 R10: 0x0000000000000000
 R11: 0x00007fd9f9fd8c00 R12: 0x0000000000000001 R13: 0x0000000000000028
 R14: 0x000055baa639cf00 R15: 0x0000000000000000 EFL: 0x0000000000010216

-- C level backtrace information -------------------------------------------
/home/watson/.rbenv/versions/3.0.0-debug/bin/ruby(rb_print_backtrace+0x19) [0x55baa0bd956f] vm_dump.c:758
/home/watson/.rbenv/versions/3.0.0-debug/bin/ruby(rb_vm_bugreport+0xb6) [0x55baa0bd99e8] vm_dump.c:998
/home/watson/.rbenv/versions/3.0.0-debug/bin/ruby(rb_bug_for_fatal_signal+0x123) [0x55baa0c6ffbe] error.c:786
/home/watson/.rbenv/versions/3.0.0-debug/bin/ruby(sigsegv+0x6d) [0x55baa0b326e0] signal.c:960
/home/watson/.rbenv/versions/3.0.0-debug/bin/ruby(sigill) (null):0
/lib/x86_64-linux-gnu/libpthread.so.0(__restore_rt+0x0) [0x7fda0213a980]
/home/watson/.rbenv/versions/3.0.0-debug/bin/ruby(ruby_stack_length+0x1c) [0x55baa09ec7f2] gc.c:5553
/home/watson/prj/rmagick/lib/RMagick2.so(rm_progress_monitor+0x31) [0x7fd9fa3b6cf1] ../../../../ext/RMagick/rmutil.c:1546
/home/watson/prj/rmagick/lib/RMagick2.so(rm_progress_monitor) (null):0
/usr/lib/libMagickCore.so.5(SetImageProgress+0x60) [0x7fd9f9e6be10] ./magick/monitor-private.h:36
/usr/lib/libMagickCore.so.5(VerticalFilter._omp_fn.1+0x5f2) [0x7fd9f9e6e6f2] ./magick/monitor-private.h:33
/usr/lib/x86_64-linux-gnu/libgomp.so.1(0x7fd9f7ccd96e) [0x7fd9f7ccd96e]
/lib/x86_64-linux-gnu/libpthread.so.0(0x76db) [0x7fda0212f6db]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7fda0147671f]
```

Maybe, this is related to 

- https://github.com/rmagick/rmagick/pull/344
- https://github.com/rmagick/rmagick/pull/468

File: ext/RMagick/rmutil.c [hunk 1]
```diff
@@ -1539,14 +1539,15 @@ rm_progress_monitor(
     VALUE rval;
     VALUE method, offset, span;
 
-// Default Ruby minimum stack size
-#define RUBY_VM_THREAD_MACHINE_STACK_SIZE_MIN (  16 * 1024 * sizeof(VALUE)) /*   64 KB or  128 KB */
-
-    // Check stack length manually instead of ruby_stack_check() for old Ruby.
-    if (ruby_stack_length(NULL) > RUBY_VM_THREAD_MACHINE_STACK_SIZE_MIN)
+    // Check running thread.
+    if (rm_current_thread_id() != rm_main_thread_id)
```
Comment (mockdeep) [inline ext/RMagick/rmutil.c:1543]: Is there any way to send this message to the main thread?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -28,7 +28,7 @@ jobs:
     timeout-minutes: 10
     strategy:
       matrix:
-        ruby-version: [2.3, 2.4, 2.5, 2.6, 2.7]
+        ruby-version: [2.3, 2.4, 2.5, 2.6, 2.7, '3.0']
```
Comment (mockdeep) [inline .github/workflows/ci.yml:31]: I guess we should probably be wrapping all of these in quotes...

File: ext/RMagick/rmutil.c (hunk 1 above)
Comment (Watson1978) [reply ext/RMagick/rmutil.c:1543]: Unfortunately, I think there is no way to do that.

Comment (Watson1978) [reply ext/RMagick/rmutil.c:1543]: If we can use private APIs, we might run monitor feature within GVL safely.

```c
    if (ruby_thread_has_gvl_p()) {
       // thread has a GVL.
       do_monitor_feature()
    }
    else if (ruby_native_thread_p()) {
       //  thread has no GVL. get
       rb_thread_call_with_gvl(do_monitor_feature, &argv);
    }
```
like, https://github.com/ruby/ruby/blob/f8555cad2d96c80fe7fc10012456e718cb81dbab/gc.c#L10271-L10278

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Watson1978) [reply .github/workflows/ci.yml:31]: Indeed.

File: ext/RMagick/rmutil.c (hunk 1 above)
Comment (mockdeep) [reply ext/RMagick/rmutil.c:1543]: @Watson1978 would ruby core allow us to make them public APIs? For now I think what you've got here is good, though.

**Pull request merged**